src/main/flight/failsafe.c

```diff
@@ -115,58 +137,57 @@ static void failsafeOnValidDataReceived(void)
 
 void failsafeUpdateState(void)
 {
-    uint8_t i;
+    bool receivingRxData = failsafeIsReceivingRxData();
+    bool armed = ARMING_FLAG(ARMED);
 
-    if (!failsafeHasTimerElapsed()) {
-        return;
+    if (receivingRxData) {
+        failsafeState.phase = FAILSAFE_IDLE;
+        failsafeState.active = false;
     }
 
-    if (!failsafeIsEnabled()) {
-        failsafeReset();
-        return;
+    switch (failsafeState.phase) {
+        case FAILSAFE_IDLE:
+            if (!receivingRxData && armed) {
+                failsafeState.phase = FAILSAFE_RX_LOSS_DETECTED;
+            }
+            break;
+
+        case FAILSAFE_RX_LOSS_DETECTED:
+            if (failsafeShouldForceLanding(armed)) {
+                // Stabilize, and set Throttle to specified level
+                failsafeActivate();
+            }
+            break;
+
+        case FAILSAFE_LANDING:
+            if (armed) {
+                failsafeApplyControlInput();
+            }
+
+            if (failsafeShouldHaveCausedLandingByNow() || !armed) {
+
+                failsafeState.phase = FAILSAFE_LANDED;
+                failsafeState.active = false;
+                mwDisarm();
+            }
+            break;
```

Comment:
Yes, but current code sets FAILSAFE_LANDED + disarms and one failsafeUpdateState() later sets PREVENT_ARMING. This may cause problems (I did not analyze rest of code ... this just seems as possible race condition that would be hard to find).